pkg/relayer/proxy/websockets/connection.go

```diff
@@ -0,0 +1,276 @@
+package websockets
+
+import (
+	"context"
+	"crypto/tls"
+	"net/http"
+	"net/url"
+	"sync/atomic"
+	"time"
+
+	"github.com/gorilla/websocket"
+
+	"github.com/pokt-network/poktroll/pkg/observable"
+	"github.com/pokt-network/poktroll/pkg/polylog"
+	"github.com/pokt-network/poktroll/pkg/relayer"
+)
+
+const (
+	// Time allowed (in seconds) to write a message to the peer.
+	writeWaitSec = 10 * time.Second
+
+	// Time allowed (in seconds) to read the next pong message from the peer.
+	pongWaitSec = 30 * time.Second
+
+	// Send pings to peer with this period.
+	// Must be less than pongWaitSec.
+	pingPeriodSec = (pongWaitSec * 9) / 10
+)
+
+// messageSource represents the source of a message in a bidirectional connection.
+// It may be either `service_backend` or `gateway`.
+type messageSource string
+
+const (
+	messageSourceServiceBackend messageSource = "service_backend"
+	messageSourceGateway        messageSource = "gateway"
+)
+
+// message represents a message sent or received in a bidirectional connection.
+// It may be a RelayRequest coming from the gateway or a raw message coming from
+// the service backend.
+type message struct {
+	// data is the message payload
+	data []byte
+
+	// source may be either `gateay` or `service_backend`
+	source messageSource
+
+	// messageType is an int returned by the gorilla/websocket package
+	messageType int
+}
+
+// connection represents a websocket connection established between the relay miner and:
+// - a gateway
+// - a service backend
+type connection struct {
+	*websocket.Conn
+
+	ctx    context.Context
+	logger polylog.Logger
+
+	// source is the source of the connection, it may be either `service_backend` or `gateway`.
+	source messageSource
+
+	// serviceID is the service ID of the service backend.
+	serviceID string
+
+	// msgChan is the channel where the messages are received by the relay miner.
+	msgChan chan<- message
+
+	// stopChan is the channel where the connection termination and errors are sent.
+	stopChan chan<- error
+
+	// stopObservable is the observable that notifies the connection to stop and errors.
+	stopObservable observable.Observable[error]
+
+	// isClosed is a flag that indicates whether the connection is closed.
+	isClosed atomic.Bool
+}
+
+// connectServiceBackend establishes a websocket connection established by the
+// relay miner client to the service backend endpoint.
+func connectServiceBackend(serviceBackendUrl *url.URL, header http.Header) (*websocket.Conn, error) {
+	// Create a new websocket dialer according to the service backend URL scheme.
+	// If the scheme is `wss`, we need to create a new dialer with a custom TLS
+	// configuration.
+	// Otherwise, we use the default dialer.
+	var dialer *websocket.Dialer
+	switch serviceBackendUrl.Scheme {
+	case "wss":
+		dialer = &websocket.Dialer{TLSClientConfig: &tls.Config{}}
+	default:
+		dialer = websocket.DefaultDialer
+	}
+
+	// TODO_TECHDEBT: Remove this switch statement when the service backend URL
```

Comment:
```suggestion
	// TODO_TECHDEBT(@red-0ne): Remove this switch statement when the service backend URL
```

Do we have a ticket for this?

Reply:
Solved